### PR TITLE
Blog post on Rust/Cryptography

### PR DESCRIPTION
**Pull request by hoodmane**
@reaperhulk @alex @davidhewitt @messense @sbc100

### CODE REVIEW

File: content/posts/rust.md [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+
+
+### Acknowledgements
+
+Thanks to the Cryptography team for making this necessary and for their
+technical advice. Thanks to the <code class="pkg">chrono</code> maintainers and
+the Rust team for reviewing and merging patches to improve Emscripten support.
+Thanks to the PyO3 team for their help and enthusiasm and for adding Emscripten
+tests to their continuous integration. Special thanks to the Emscripten team,
+particularly Sam Clegg, for their technical advice, for merging patches, and for
```
Comment (alex) [inline content/posts/rust.md:71]: Wowa, small world! Long time no see Sam!

**Review by alex — COMMENTED** [review]
Looks great to me. Thanks for diving into all of these issues and I'm glad its mostly working now!

Hopefully someday soon we'll be able to drop Python 3.6 and remove the need to patch `instant` and `chrono`.

File: content/posts/rust.md (hunk 1 above)
Comment (sbc100) [reply content/posts/rust.md:71]: Nice!  Great to see you name pop up in my feed Alex.  I hope you are good!

**Comment by messense** [comment]
The post content looks great, but the excerpt from the list of posts looks odd:

image

**Comment by reaperhulk** [comment]
Thanks for the ping, this was a fascinating read. Glad to see you found a path and I hope we'll be able to drop 3.6 early next year to improve our deps a bit!

File: content/posts/rust.md [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+ShowPostNavLinks: true
+cover:
+    image: "<image path/url>" # image path/url
+    alt: "<alt text>" # alt text
+    caption: "<text>" # display caption under cover
+    relative: false # when using page bundles set this to true
+    hidden: true # only hide on current single page
+---
+
+We are happy to announce that the next version of Pyodide will ship with the
```
Comment (ryanking13) [inline content/posts/rust.md:28]: Perhaps we can release this post along with the next Pyodide release?

**Review by ryanking13 — COMMENTED** [review]
I wasn't following your Cryptography work in detail, so I was quite surprised that there was so many issues that I didn't know. Thanks a lot again Hood, it was really fun to read.

File: content/posts/rust.md [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+newest version of the Cryptography package, including its Rust extension module.
+The Cryptography package was one of the first major packages that included a
+Rust extension module. By including Rust, the Cryptography maintainers have
+forced the Python ecosystem to adopt Rust support. We are no exception. The
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/rust.md:32]: Yeah that sounds better.

File: content/posts/rust.md [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+call.
+
+These `dynCall` functions are only generated for code in the main WebAssembly
+executable, not for code in dynamically linked libraries. If a function called
+inside of a C++ `try` block has a different signature from all functions called
+in try blocks in the main executable, we see the error above.
+
+It would be possible to generate extra `dynCall` wrappers for dynamic libraries,
+either by including them in the dynamic library or by generating them at load
+time. However, Emscripten does not support this.
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/rust.md:290]: I don't think they are likely to support it, but that's okay.

File: content/posts/rust.md [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+## Misencoded object files and LLVM version conflicts
+
+Once the `lib.rmeta` problem was resolved, the second linker error I encountered
+was that `compiler_builtins-<...>-cgu1.rcgu.o` had encoding errors. The solution
+to this was... I waited a few months, and the error went away.
+
+I believe the problem was that the Emscripten linker uses a different version of
+LLVM than the Rust compiler and the object file format was slightly different in
+the two LLVM versions. According to this theory, the problem went away because
+Rust updated LLVM. We're still using a different LLVM versions to compile and to
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/rust.md:168]: Well I think the problem was present in Rust 1.58 and went away either in Rust 1.59 released on Feb 22 or in 1.60 published on April 4th.

File: content/posts/rust.md [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+more stable than the Emscripten shared library format so this would be worse
+than our current approach.
+
+## Unable to find library `-lc-debug`
+
+The next linker error is
+```
+Unable to find library `-lc-debug`
+```
+For some reason, Rust tries to link `libc` into Emscripten dynamic libraries. It
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/rust.md:192]: Agreed.
See also discussion:
https://github.com/emscripten-core/emscripten/issues/17191#issuecomment-1151607689
https://github.com/rust-lang/rust/pull/98303

**Comment by hoodmane** [comment]
> it's already a long post and it touches on a lot of things, so making it even longer would not be ideal.

I would also be fine with removing some of the less interesting sections. But in case anyone else trying to build Rust code for `wasm32-unknown-emscripten` hits similar errors, I thought it might be helpful to be thorough.

File: content/posts/rust.md [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+
+See [this talk by the Cryptography
+maintainers](https://www.youtube.com/watch?v=z_Eiy2W0APU) for an interesting
+discussion of their reasons for using Rust and the problems in the ecosystem
+that needed to be fixed before they could use it.
+
+We want to build the Rust extension module for our WebAssembly-based platform.
+Rust has good support for the <code class="pkg">wasm32-unknown-unknown</code>
+target and has popular tools like <code class="pkg">wasm-bindgen</code>.
+However, we need to use the <code class="pkg">wasm32-unknown-emscripten</code>
```
[the remark this answers is not in the document]
Comment (hoodmane) [reply content/posts/rust.md]: Updated it to say:

> Rust has good support for the wasm32-unknown-unknown target and has popular tools like wasm-bindgen. However, wasm32-unknown-unknown uses a custom “wasm ABI” [which is not compatible with C/C++ code](https://github.com/rustwasm/team/issues/291#issuecomment-644946504l). To use Rust in a project that also includes C/C++ code we need to use the wasm32-unknown-emscripten compiler target.

File: content/posts/rust.md (hunk 2 above)
Comment (ryanking13) [reply content/posts/rust.md]: > to create some anticipation :)

Sounds good to me :)

File: content/posts/rust-pyo3-support-in-pyodide.md [hunk 8]
```diff
@@ -0,0 +1,338 @@
...
+code](https://github.com/rustwasm/team/issues/291#issuecomment-644946504l). To
+use Rust in a project that also includes C/C++ code we need to use the <code
+class="pkg">wasm32-unknown-emscripten</code> target. Unfortunately, the <code
+class="pkg">wasm32-unknown-emscripten</code> target does not work all that well,
+due to limited development resources and difficulties coordinating between the
+Emscripten and Rust projects.
+
+
+In this blog post I will give a technical description of many of the problems I
+ran into in the process of building Rust extension modules for Pyodide.
```
Comment (davidhewitt) [inline content/posts/rust-pyo3-support-in-pyodide.md:57]: Maybe it's nice to be clear that you did manage to solve them all and readers will learn about the solutions too?

```suggestion
In this blog post I will give a technical description of many of the challenges I
ran into in the process of building Rust extension modules for Pyodide and their solutions.
```

File: content/posts/rust-pyo3-support-in-pyodide.md [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+  margin: 0px;
+}
+</style>
+
+### Acknowledgements
+
+Thanks to the Cryptography team for making this necessary and for their
+technical advice. Thanks to the <code class="pkg">chrono</code> maintainers and
+the Rust team for reviewing and merging patches to improve Emscripten support.
+Thanks to the PyO3 team for their help and enthusiasm and for adding Emscripten
```
Comment (davidhewitt) [inline content/posts/rust-pyo3-support-in-pyodide.md:77]: 🙌

File: content/posts/rust-pyo3-support-in-pyodide.md [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+export somehow.
+
+Using `-sSIDE_MODULE=2` solves the `lib.rmeta` error because `--whole-archive`
+is not passed to the linker. Conveniently, Rust is good at calculating which
+symbols should be exported: public symbols with the `#[no_mangle]` attribute are
+exported, other symbols are not. In the case of a PyO3 module, the only exported
+symbol is the `PyInit__my_module` function that Python invokes when loading a
+native module. 
+
+It would be great to link our other extension modules with `-sSIDE_MODULE=2` but
```
Comment (davidhewitt) [inline content/posts/rust-pyo3-support-in-pyodide.md:165]: I'm a little unsure what this is aspiring to (what "our other extension modules" are).

File: content/posts/rust-pyo3-support-in-pyodide.md [hunk 11]
```diff
@@ -0,0 +1,338 @@
...
+Building with `-sSIDE_MODULE=2` is much better for code size because symbols
+that are not exported can be inlined or eliminated as dead code during link-time
+optimization. But to use `-sSIDE_MODULE=2` we need to calculate the symbols to
+export somehow.
+
+Using `-sSIDE_MODULE=2` solves the `lib.rmeta` error because `--whole-archive`
+is not passed to the linker. Conveniently, Rust is good at calculating which
+symbols should be exported: public symbols with the `#[no_mangle]` attribute are
+exported, other symbols are not. In the case of a PyO3 module, the only exported
+symbol is the `PyInit__my_module` function that Python invokes when loading a
```
Comment (davidhewitt) [inline content/posts/rust-pyo3-support-in-pyodide.md:162]: So does Rust pass this symbol automatically to the linker, or did you have to add it to the command line?

**Review by davidhewitt — COMMENTED** [review]
Sorry to have been a bit slow to give this a read. Thanks for writing this, really interesting to see the whole picture, and thanks also for working on the wider problem space & Pyodide!

Had a few quick thoughts, feel free to simply ignore them ofc 😄

File: content/posts/rust-pyo3-support-in-pyodide.md (hunk 11 above)
Comment (hoodmane) [reply content/posts/rust-pyo3-support-in-pyodide.md:162]: Rust automatically passes the information to the linker when building a cdylib, this is handled here:
https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_ssa/src/back/link.rs#L1882

**Comment by rth** [comment]
@hoodmane Feel free to merge once you have responded to above comments. Just need to make sure that the date in the header is today and in the past (otherwise if it's in the future it won't be published).

Also if you have a short description for posting this on Twitter, please send it to me, otherwise, I'll write something. Thanks!

**Pull request merged**